snap/validate.go

```diff
@@ -71,6 +73,92 @@ func ValidateAlias(alias string) error {
 	return nil
 }
 
+// ValidateAppSocketName checks if a string ca be used as a name for a socket
+// (for socket activation).
+func ValidateAppSocketName(name string) error {
+	valid := validSnapName.MatchString(name)
+	if !valid {
+		return fmt.Errorf("invalid socket name: %q", name)
+	}
+	return nil
+}
+
+// ValidateAppSocketListenAddress checks that the value of socket addresses.
+func ValidateAppSocketListenAddress(socket *SocketInfo, fieldName string, address string) error {
+	if address == "" {
+		return fmt.Errorf("socket %q must define %q", socket.Name, fieldName)
+	}
+
+	switch address[0] {
+	case '/', '$':
+		return validateAppSocketListenAddressPath(socket, fieldName, address)
+	case '@':
+		return validateAppSocketListenAddressAbstractSocket(socket, fieldName, address)
+	default:
+		return validateAppSocketListenAddressNetAddress(socket, fieldName, address)
+	}
+}
+
+func validateAppSocketListenAddressPath(socket *SocketInfo, fieldName string, path string) error {
+	if path != filepath.Clean(path) {
+		return fmt.Errorf(
+			`socket %q has invalid %q: paths must not include "." or ".."`, socket.Name, fieldName)
+	}
+
+	if !(strings.HasPrefix(path, "$SNAP_DATA/") || strings.HasPrefix(path, "$SNAP_COMMON/")) {
+		return fmt.Errorf(
+			"socket %q has invalid %q: only $SNAP_DATA and $SNAP_COMMON prefixes are allowed", socket.Name, fieldName)
+	}
+
+	return nil
+}
+
+func validateAppSocketListenAddressAbstractSocket(socket *SocketInfo, fieldName string, path string) error {
+	prefix := "@snap." + socket.App.Snap.Name()
+	if !strings.HasPrefix(path, prefix) {
+		return fmt.Errorf("socket %q path for %q must be prefixed with %q", socket.Name, fieldName, prefix)
+	}
+	return nil
+}
+
+func validateAppSocketListenAddressNetAddress(socket *SocketInfo, fieldName string, address string) error {
+	lastIndex := strings.LastIndex(address, ":")
+	if lastIndex >= 0 {
+		if err := validateAppSocketListenAddressNetAddressAddress(socket, fieldName, address[:lastIndex]); err != nil {
+			return err
+		}
+		if err := validateAppSocketListenAddressNetAddressPort(socket, fieldName, address[lastIndex+1:]); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// Address only contains a port
+	if err := validateAppSocketListenAddressNetAddressPort(socket, fieldName, address); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateAppSocketListenAddressNetAddressAddress(socket *SocketInfo, fieldName string, address string) error {
+	for _, validAddress := range []string{"127.0.0.1", "[::1]", "[::]"} {
```

Comment:
not `"localhost"`?

Reply:
Is that supported by systemd? the manpage only mentions IP addresses

Reply:
no! sorry